activation/nipost_test.go

```diff
@@ -41,6 +48,337 @@ func defaultLayerClockMock(ctrl *gomock.Controller) *MocklayerClock {
+		require.NotNil(t, nipost)
+		require.NotNil(t, nipostInfo)
+	})
+
+	t.Run("connect, disconnect, reconnect then error", func(t *testing.T) {
+		// post client connects, starts post, disconnects in between and then fails to complete
+
+		sig, err := signing.NewEdSigner()
+		require.NoError(t, err)
+
+		ctrl := gomock.NewController(t)
+		poetDb := NewMockpoetDbAPI(ctrl)
+		mclock := defaultLayerClockMock(ctrl)
+		postService := NewMockpostService(ctrl)
+		postClient := NewMockPostClient(ctrl)
+
+		logger := zaptest.NewLogger(t)
+		nb, err := NewNIPostBuilder(
+			localsql.InMemory(),
+			poetDb,
+			postService,
+			[]types.PoetServer{},
+			logger,
+			sig,
+			PoetConfig{},
+			mclock,
+		)
+		require.NoError(t, err)
+
+		expectedEvents := setupEventListener(t)
+		expectedErr := errors.New("some error")
+
+		postService.EXPECT().Client(sig.NodeID()).DoAndReturn(
+			func(types.NodeID) (PostClient, error) {
+				expectedEvents <- func(e events.UserEvent) {
+					event := e.Event.GetPostStart()
+					if !assert.NotNil(t, event, "wrong event type") {
+						return
+					}
+
+					assert.EqualValues(t, shared.ZeroChallenge, event.Challenge)
+				}
+				return postClient, nil
+			})
+		postClient.EXPECT().Proof(gomock.Any(), gomock.Any()).Return(nil, nil, ErrPostClientClosed).Times(1)
+		postService.EXPECT().Client(sig.NodeID()).Return(postClient, nil)
+		postClient.EXPECT().Proof(gomock.Any(), gomock.Any()).DoAndReturn(
+			func(context.Context, []byte) (*types.Post, *types.PostInfo, error) {
+				expectedEvents <- func(e events.UserEvent) {
+					event := e.Event.GetPostComplete()
+					if !assert.NotNil(t, event, "wrong event type") {
+						return
+					}
+
+					assert.Equal(t, true, e.Event.Failure)
+					assert.Equal(t, "Node failed PoST execution.", e.Event.Help)
+				}
+
+				return nil, nil, expectedErr
+			})
+
+		nipost, nipostInfo, err := nb.Proof(context.Background(), shared.ZeroChallenge)
+		require.ErrorIs(t, err, expectedErr)
+		require.Nil(t, nipost)
+		require.Nil(t, nipostInfo)
+	})
+
+	t.Run("repeated connection failure", func(t *testing.T) {
+		t.Parallel()
+
+		sig, err := signing.NewEdSigner()
+		require.NoError(t, err)
+
+		ctrl := gomock.NewController(t)
+		poetDb := NewMockpoetDbAPI(ctrl)
+		mclock := defaultLayerClockMock(ctrl)
+		postService := NewMockpostService(ctrl)
+
+		observer, observedLogs := observer.New(zapcore.WarnLevel)
+		logger := zap.New(observer)
+
+		nb, err := NewNIPostBuilder(
+			localsql.InMemory(),
+			poetDb,
+			postService,
+			[]types.PoetServer{},
+			logger,
+			sig,
+			PoetConfig{},
+			mclock,
+		)
+		require.NoError(t, err)
+
+		setupEventListener(t) // no events
+
+		ctx, cancel := context.WithCancel(context.Background())
+		postService.EXPECT().Client(sig.NodeID()).Return(nil, ErrPostClientNotConnected).Times(10)
+		postService.EXPECT().Client(sig.NodeID()).DoAndReturn(
+			func(types.NodeID) (PostClient, error) {
+				cancel()
+				return nil, ErrPostClientNotConnected
+			}).Times(1)
```

Comment:
`Times(1)` is implicit
```suggestion
			})
```